packages/client-node/src/result_set.ts

```diff
@@ -35,34 +38,64 @@ export class ResultSet implements BaseResultSet<Stream.Readable> {
 
     validateStreamFormat(this.format)
 
-    let decodedChunk = ''
+    let data: Buffer
+    let incompleteRow: Buffer | undefined
     const toRows = new Transform({
       transform(
         chunk: Buffer,
-        encoding: BufferEncoding,
+        _encoding: BufferEncoding,
         callback: TransformCallback
       ) {
-        decodedChunk += chunk.toString()
+        let currentDataSize
+        // The buffer was already allocated
+        // x2 chunk size should be enough for any leftover incomplete rows
+        if (data !== undefined) {
+          let chunkOffset
+          if (incompleteRow !== undefined) {
+            chunkOffset = incompleteRow.length
+            currentDataSize = chunkOffset + chunk.length
+            data.fill(incompleteRow, 0, chunkOffset)
+          } else {
+            chunkOffset = 0
+            currentDataSize = chunk.length
+          }
+          data.fill(chunk, chunkOffset, currentDataSize)
+          data.fill(0, currentDataSize, data.length)
+        } else {
+          // This is the first chunk, and we need to allocate enough memory
+          // to hold the chunk itself and any potential incomplete rows
+          currentDataSize = chunk.length
+          data = Buffer.alloc(chunk.length * 2, 0)
```

Comment:
Surprisingly, `allocUnsafe` without additional zero fills does not have any benefits here, at least not on the `hits_v1` dataset which is rather large (~9M rows).